### PR TITLE
feat(glm53): add qualified TP3 support to R21

### PR DESCRIPTION
**Pull request by infernix**
## Summary

Adds GLM-5.3-Flash physical TP3/DCP1/EP3 support on the exact R21 generic-container source baseline while retaining the existing TP4 path.

- validates physical TP3 padding for attention, GDN, routed and shared experts, vocabulary, MTP, DFlash2, and multimodal weights mode
- supports ordinary, MTP3, and DFlash2 K7 serving
- keeps TP1, TP2, TP4, and unrelated model configurations unchanged
- rejects unsupported TP3 configurations before model construction
- preserves RoCE fallback, shutdown, pooled-indexer, and speculative-topology behavior

## Source identity

- base branch: `artifact/jovian-judgement-community-20260904-r21-source`
- exact base: `f2d77086163e899f87f54a59af216d18ffa3a2b7`
- head: `e96b18dbb8c19230591e79e0ed056b12947b2ea1`
- tree: `31e73a43eb8a03e932f03c51341df2c73c60f3d4`
- R21 parent: `voipmonitor/vllm:jovian-judgement-community-20260904-r21@sha256:f096012c508f9bc12e8c4e617b8ed19da3a2cecb525e9479904e848730f0c8ac`

## Compatibility

The TP3 path is selected only for the validated GLM-5.3 physical geometry and fails closed for incompatible topology, backend, checkpoint, or multimodal settings. Existing non-TP3 dispatch remains unchanged.

## Verification

Source gates on the R21 image:

- TP3 geometry: **28 passed**
- GLM model: **19 passed**
- DFlash: **9 passed**
- KDA stream: **3 passed, 2 skipped**
- pooled indexer: **13 passed, 31 skipped**
- changed-source byte compilation: **passed**

## Hardware qualification

The published R21 child image is pull-ready for end users without rebuilding:

- immutable manifest digest: `infernix/vllm@sha256:e81f9399aa9fe800593cc8f646d8a2c7958e1938da50c5ae65effbe47d8604eb`
- secondary convenience tag: `infernix/vllm:glm53-r21-tp3-qualified-vllme96b18dbb8c1-b12x6d47b10eddf4-recipe01c67936a364`
- qualified image config ID: `sha256:dbdc64cb31b0c2bc1a0bbd5eaa4e5a91a1539333123547cc6de9b08c426bf6c1`

The image with that config ID passed on four NVIDIA RTX PRO 6000 Blackwell Workstation Edition GPUs:

- TP3 ordinary: health, startup, and exact text smoke
- TP3 MTP3: health, startup, and exact text smoke
- TP3 DFlash2 K7: health, startup, exact text, red-image vision, 8/8 concurrent requests, and 1,000,031-prompt-token retrieval
- TP4 MTP3 regression against the pinned target revision: health, startup, and exact text smoke

## Related work

- B12X integration: local-inference-lab/b12x#265
- source-locked R21 runtime overlay: local-inference-lab/blackwell-llm-docker#30
- qualification receipt: local-inference-lab/rtx6kpro#92

### CODE REVIEW

File: vllm/models/minimax_m3/common/indexer.py [hunk 1]
```diff
@@ -385,13 +429,45 @@ def forward(
...
+        raise NotImplementedError
+
+    def _topk_out_buffer(
+        self,
+        kind: str,
+        total_q: int,
+    ) -> torch.Tensor | None:
+        buffer = (
+            self._decode_topk_buffer if kind == "decode" else self._prefill_topk_buffer
+        )
```
[the remark this answers is not in the document]
Comment (infernix) [reply vllm/models/minimax_m3/common/indexer.py]: Fixed in 9fe966ebc. The base explicit-cache path now temporarily binds the supplied cache, dispatches through the implementation-specific `forward`, and restores the owned cache in `finally`, so the MSA implementation uses its own forward path instead of inheriting NotImplementedError.

File: vllm/model_executor/layers/linear.py [hunk 2]
```diff
@@ -1229,6 +1284,35 @@ def weight_loader(
...
+                            )
+                            * self.head_size
+                            + self.loaded_total_num_kv_heads * self.v_head_size,
+                            0,
+                        ),
+                    }
+
+                    shard_size, shard_offset = adjust_bitsandbytes_4bit_shard(
+                        param, orig_qkv_offsets, shard_id
+                    )
```
[the remark this answers is not in the document]
Comment (infernix) [reply vllm/model_executor/layers/linear.py]: Fixed in 9fe966ebc. The helper and core BnB loader path were removed when BitsAndBytes moved to the plugin; only this orphaned call site had been replayed. The obsolete branch is removed, so fused QKV checkpoints use the supported slicing path and cannot raise NameError.

File: vllm/model_executor/models/qwen3_dflash2.py [hunk 3]
```diff
@@ -258,7 +258,7 @@ def __init__(
         with set_model_tag("dflash2_candidate_selector"):
             self.candidate_selector = CandidateSelector(
                 hidden_size=self.config.hidden_size,
-                vocab_size=self.config.vocab_size,
+                vocab_size=self.config.draft_vocab_size,
```
[the remark this answers is not in the document]
Comment (infernix) [reply vllm/model_executor/models/qwen3_dflash2.py]: Fixed in 9fe966ebc. CandidateSelector now uses the target config's logical vocabulary via `_get_dflash_draft_vocab_size(vllm_config.model_config.hf_config)`. That covers target-head/input IDs while preserving TP3's 154880 logical domain instead of incorrectly exposing the 154944 padded storage tail; TP4 remains the 154880 no-op case.

File: vllm/models/minimax_m3/common/indexer.py [hunk 4]
```diff
@@ -427,6 +503,7 @@ def forward(
                 self.init_blocks,
                 self.local_blocks,
                 self.num_kv_heads,
+                self.scale,
```
[the remark this answers is not in the document]
Comment (infernix) [reply vllm/models/minimax_m3/common/indexer.py]: Fixed in 9fe966ebc. Removed `self.scale` from both kernel calls; the remaining positional arguments now align with the NVIDIA and AMD decode/score signatures and `out` is no longer misbound.

File: vllm/models/glm5next/nvidia/model.py [hunk 5]
```diff
@@ -266,6 +287,11 @@ def __init__(
                 reduce_results=False,
                 prefix=f"{prefix}.shared_experts",
                 swiglu_limit=swiglu_limit,
+                loaded_intermediate_size=loaded_intermediate_size,
+                disable_tp=replicate_shared_expert,
+                output_scale=(
+                    1.0 / self.tp_size if replicate_shared_expert else 1.0
+                ),
```
[the remark this answers is not in the document]
Comment (infernix) [reply vllm/models/glm5next/nvidia/model.py]: Fixed in `73e6bd12c7`. Replicated shared experts now use `1.0` under sequence parallelism and apply `1.0 / tp_size` only on the non-sequence-parallel path. A parameterized constructor regression covers both cases.

File: vllm/v1/core/sched/scheduler.py [hunk 6]
```diff
@@ -1898,8 +1898,13 @@ def update_from_output(
...
                 if self.acceptance_length_controller is not None:
+                    num_invalid_tokens = (
+                        scheduler_output.num_invalid_spec_tokens or {}
+                    ).get(req_id, 0)
                     adaptive_num_drafts += 1
-                    adaptive_num_draft_tokens += num_draft_tokens
+                    adaptive_num_draft_tokens += (
+                        num_draft_tokens - num_invalid_tokens
+                    )
                     adaptive_num_accepted_tokens += num_accepted
```
[the remark this answers is not in the document]
Comment (infernix) [reply vllm/v1/core/sched/scheduler.py:2513]: Fixed in `73e6bd12c7`. Adaptive accepted/draft counters now exclude stale output before `observe_batch`; a regression marks a verify result stale and confirms the controller receives zero fresh observations for that batch.

**Comment by infernix** [comment]
@coderabbitai review

**Comment by infernix** [comment]
@coderabbitai review

File: vllm/distributed/device_communicators/b12x_pcie_all_reduce.py [hunk 7]
```diff
@@ -47,6 +47,29 @@ def _parse_byte_size(value: str) -> int:
...
 
 
+def _twoshot_max_bytes() -> int:
+    """Largest all-reduce routed to the lossless bf16 two-shot (0 disables).
+
+    Measured on four RTX PRO 6000 Blackwell (PCIe, TP4): the pull two-shot
+    beats the NCCL ring from the one-shot ceiling up to 768 KB (128 KB 15.0
+    vs 16.7 us, 512 KB 32.8 vs 41.7, 768 KB 44.5 vs 54.5) and matches it at
+    1 MB and above, so the default stops at 768 KB.
+    """
```
[the remark this answers is not in the document]
Comment (infernix) [reply vllm/distributed/device_communicators/b12x_pcie_all_reduce.py:57]: No runtime change. The missing Returns section is documentation-only inside the installed qualified package tree; changing it would sever the exact image/package identity without changing behavior. The return type and value are already explicit (int byte limit).

File: vllm/model_executor/models/qwen3_dflash.py [hunk 8]
```diff
@@ -101,6 +101,41 @@ def dflash_target_rope_is_neox_style(target_model: nn.Module) -> bool | None:
             return style
     return None
 
+def _get_dflash_draft_vocab_size(config: Qwen3Config) -> int:
+    """Return the logical draft vocabulary across old and current config APIs."""
+    draft_vocab_size = getattr(config, "draft_vocab_size", None)
+    if draft_vocab_size is None:
```
[the remark this answers is not in the document]
Comment (infernix) [reply vllm/model_executor/models/qwen3_dflash.py:107]: No qualified-path change. The locked draft has vocab_size=154880 and no draft_vocab_size; TP3 additionally sets original_vocab_size and draft_vocab_size to 154880 before construction. Explicit zero is malformed and unreachable in both qualified layouts.

File: vllm/distributed/device_communicators/b12x_pcie_all_reduce.py [hunk 9]
```diff
@@ -272,6 +299,57 @@ def _initialize_dma(self, dma_cls: Any) -> None:
...
+    def _initialize_twoshot(self) -> None:
+        """Lossless bf16 two-shot for payloads above the one-shot ceiling."""
+        max_bytes = _twoshot_max_bytes()
+        if max_bytes <= 0 or self.world_size not in (2, 4, 8):
+            return
+        twoshot_cls = _load_b12x_twoshot_bf16()
+        if twoshot_cls is None:
+            logger.warning("B12X PCIe two-shot bf16 requested but unavailable.")
+            return
+        row_elems = int(os.getenv("VLLM_PCIE_TWOSHOT_ROW_ELEMS", "4096"))
```
[the remark this answers is not in the document]
Comment (infernix) [reply vllm/distributed/device_communicators/b12x_pcie_all_reduce.py:311]: No qualified-path change. TP3 exits two-shot initialization before reading this variable; qualified TP4 uses the valid built-in value 4096 and the locked launcher does not set VLLM_PCIE_TWOSHOT_ROW_ELEMS. Malformed undocumented operator overrides remain fail-fast rather than a coordinated fallback contract.

File: tests/models/test_glm53_tp3_model.py [hunk 10]
```diff
@@ -0,0 +1,623 @@
...
+        output_size=1,
+        params_dtype=torch.bfloat16,
+        weight_loader=lambda *_args, **_kwargs: None,
+    )
+
+    assert holder.weight_scale.input_dim_storage_factor == 32
+    loaded = torch.arange(1, 129, dtype=holder.weight_scale.dtype).unsqueeze(0)
+    row = RowParallelLinear(12288, 1, bias=False, loaded_input_size=4096)
+    row.weight_loader_v2(holder.weight_scale, loaded)
+    torch.testing.assert_close(holder.weight_scale, torch.zeros_like(holder.weight_scale))
```
[the remark this answers is not in the document]
Comment (infernix) [reply tests/models/test_glm53_tp3_model.py]: Fixed in 059f85d258. Ruff reformatted both over-length test lines; focused ruff format/check now pass.

File: vllm/model_executor/layers/linear.py [hunk 11]
```diff
@@ -39,12 +39,123 @@
...
+                    f"{name} {boundary_name} {boundary} is not aligned to "
+                    f"packed_factor={packed_factor}"
+                )
+        if isinstance(param, (PackedColumnParameter, PackedvLLMParameter)):
+            packed_size, _ = param.adjust_shard_indexes_for_packing(
+                shard_size=loaded_size, shard_offset=0
+            )
+            return packed_size
+
+    return loaded_size
```
[the remark this answers is not in the document]
Comment (infernix) [reply vllm/model_executor/layers/linear.py:156]: No qualified-path change. The mismatch is real for generic FPQuant padded loading, but the qualified target is ModelOpt mixed NVFP4/MXFP8 and the DFlash draft is ModelOpt MXFP8; neither constructs FPQuantLinearMethod parameters. This needs a separate generic FPQuant compatibility change and qualification.

File: vllm/v1/attention/backends/dflash_attn/__init__.py [hunk 12]
```diff
@@ -0,0 +1,256 @@
...
+        bf16 views with a unit last stride (K and V identically strided);
+        block_table [B, max_blocks] int32; seqused_k [B] int32; cu [B+1] int32."""
+        batch = int(cu_seqlens_q.numel() - 1) if num_reqs is None else int(num_reqs)
+        if batch > self.max_batch:
+            raise ValueError(f"batch {batch} exceeds workspace {self.max_batch}")
+        heads = q.shape[1]
+        if heads != self.hkv * GQA or q.shape[2] != HEAD_DIM:
+            raise ValueError("unsupported head geometry")
+        if not (q.is_contiguous() and out.is_contiguous()):
+            raise ValueError("q and out must be contiguous")
```
[the remark this answers is not in the document]
Comment (infernix) [reply vllm/v1/attention/backends/dflash_attn/__init__.py:204]: No qualified-path change. The custom split-KV CUDA kernel is opt-in through VLLM_GLM53_DFLASH_ATTN=1 and is not enabled by the qualified recipe, which records FlashAttention 2. Qualified query/output are also BF16. Mixed FP16 query/output support belongs with separately enabling and qualifying this optional kernel.

File: vllm/v1/attention/backends/mla/b12x_mla_sparse.py [hunk 13]
```diff
@@ -1049,26 +1130,62 @@ def __init__(
...
+                "nvfp4_ds_mla KV cache; "
                 f"got kv_cache_dtype={kv_cache_dtype!r}."
             )
+        self._uses_nvfp4_cache = uses_nvfp4_cache
+        if self._is_glm_next and self._uses_nvfp4_cache:
+            self._cache_record_bytes = _GLM_NEXT_NVFP4_CACHE_RECORD_BYTES
+        elif self._uses_glm_dsa_nvfp4_cache:
+            self._cache_record_bytes = _GLM_DSA_NVFP4_CACHE_RECORD_BYTES
+        else:
+            self._cache_record_bytes = _GLM_NEXT_CACHE_RECORD_BYTES
```
[the remark this answers is not in the document]
Comment (infernix) [reply vllm/v1/attention/backends/mla/b12x_mla_sparse.py:1172]: No qualified-path change. The 656-byte generic fp8_ds_mla record is relevant to non-GLM5Next models. The locked target is glm5_next_text, whose customized cache spec and implementation both intentionally use 528 bytes. A generic DeepSeek sparse-MLA correction should be separate.

File: tests/v1/attention/test_dflash_attn.py [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+
+
+def test_workspace_rejects_oversized_batch():
+    dfa = _module()
+    device = torch.device("cuda", 0)
+    q, k, v, block_table, cu, seqused = _case([64, 64], [8, 8], True, device)
+    op = dfa.DFlashDecodeAttention(device, HKV, max_batch=1, window=WINDOW)
+    with pytest.raises(ValueError):
+        op(q, k, v, block_table, seqused, cu, SCALE, torch.empty_like(q))
+    assert os.getenv("VLLM_GLM53_DFLASH_ATTN", "0") in ("0", "1")
```
[the remark this answers is not in the document]
Comment (infernix) [reply tests/v1/attention/test_dflash_attn.py]: Fixed in 059f85d258. Removed the ambient VLLM_GLM53_DFLASH_ATTN assertion and now-unused os import; the test only checks the oversized-workspace contract.

**Comment by infernix** [comment]
@coderabbitai review

**Comment by infernix** [comment]
@coderabbitai review
